### ISSUE TITLE
ZMQ died silently

### ISSUE THREAD

**Issue by engenegr** [ACINQ/eclair#1895]
ZMQ silently died again without signal about disconnecting with node.
Last signal I got on my node was from
```
2021-08-02 10:27:36,964 INFO  f.a.e.b.b.zmq.ZMQActor - connected to tcp://192.168.2.7:29000
```
then I used it for a while and on 6th Aug node didn't signal about ZMQ disruptions while Watchdog indicated eclipsing.

```
2021-08-06 12:39:20,703 WARN  f.a.e.b.w.BlockchainWatchdog - blockcypher.com: we are 10 blocks late: we may be eclipsed from the bitcoin network
2021-08-06 12:54:10,281 WARN  f.a.e.b.w.BlockchainWatchdog - blockstream.info: we are 13 blocks late: we may be eclipsed from the bitcoin network
2021-08-06 12:54:10,998 WARN  f.a.e.b.w.BlockchainWatchdog - mempool.space: we are 13 blocks late: we may be eclipsed from the bitcoin network
2021-08-06 12:54:11,860 WARN  f.a.e.b.w.BlockchainWatchdog - bitcoinheaders.net: we are 9 blocks late: we may be eclipsed from the bitcoin network
2021-08-06 12:54:20,448 WARN  f.a.e.b.w.BlockchainWatchdog - blockcypher.com: we are 13 blocks late: we may be eclipsed from the bitcoin network
```
Restarting BITCOIND didn't help that's why I think ZMQ actor died.
version `0.6.2-SNAPSHOT-cafaeed`

**Comment by btcontract** [comment]
> on 6th Aug node didn't signal about ZMQ disruptions while Watchdog indicated eclipsing

I assume this is about alarmbot plugin not sending a message to telegram? 

That would be strange as dangerous skew event is present in logs. So it could be that plugin actor did not receive a dangerous skew message for some reason (as strange and unlikely as it sounds given that subscription is straightforward). Guess plugin should be tested more thoroughly here.

**Comment by engenegr** [comment]
> > on 6th Aug node didn't signal about ZMQ disruptions while Watchdog indicated eclipsing
> 
> I assume this is about alarmbot plugin not sending a message to telegram?
> 

Yes. Alarmbot also cant signal about Dangerous Skew. Node running and channel opening messages work as expected. 

Actually, I never received Dangerous Skew messages. To me, it has somewhat lower priority now. I should forget about ZMQ issues first to really care about this issue.

**Comment by t-bast** [comment]
Did you check that your `bitcoind` node did have newer blocks that weren't sent to eclair on the ZMQ socket?
It could be that your `bitcoind` node was indeed being eclipsed (or didn't receive blocks) for some reason?
Could you activate `debug` logs for the `ZmqActor`? That would let us see whether it receives data or not.

**Comment by engenegr** [comment]
@t-bast 
hey, yes, I've checked the node. Then I tried to restart it even. 

Sure I could try enable logs but I also report about that there were no messages about disconnected ZMQ as well when I restarted Core. Usually Eclair sends message about ZMQ disconnect, that's right.

**Comment by t-bast** [comment]
We need the debug logs to have *some* information, otherwise we're completely in the dark to investigate...

> I also report about that there were no messages about disconnected ZMQ as well when I restarted Core.

You mean that you restarted the bitcoin core node, and that didn't cause any logs from the zmq actor, not even to advertise the disconnect/reconnect? If the restart was fast, maybe that's handled by the zmq socket transparently, but that would mean it stayed connected, so in that case we'd need to understand why it didn't receive messages if it was correctly connected...

**Comment by engenegr** [comment]
@t-bast right. Your comments are perfectly legit. However, it may be hard to switch on verbose logs and repeat exactly the same situation.

Just today I could closer observe similar situation since it was in 2-hour timespan or something.

My node critically failed. I do not know what really happened but Linux just blacked out. Eclair didn't signalled about the failure. I restarted the node, and Core found itself late about block-day (> 144 blocks). During several blocks going forward Eclair keep sending Watchdog eclipsing message. Only after dozens of minutes I got ZMQ - disconnect  ZMQ-connect messages one after another. However, Core was still late to the tip. While Core was syncing Eclair sent same message one more time.

```
2021-08-11 14:11:17,875 WARN  f.a.e.b.b.zmq.ZMQActor - disconnected from tcp://192.168.2.7:29000
2021-08-11 14:11:19,905 INFO  f.a.e.b.b.zmq.ZMQActor - connected to tcp://192.168.2.7:29000
2021-08-11 14:15:44,074 WARN  f.a.e.b.b.zmq.ZMQActor - disconnected from tcp://192.168.2.7:29000
2021-08-11 14:15:48,154 INFO  f.a.e.b.b.zmq.ZMQActor - connected to tcp://192.168.2.7:29000
```

Finally, Core synced and Eclair just started to show tip height in `getinfo`. 
That's it for now. Sorry it may be not really good report but this is what I have.

**Comment by t-bast** [comment]
Thanks for the additional details.

I think it would be useful if you could turn on debug logs for the `ZMQActor`, if a similar issue happens in the future we will then have more data to work with. You can enable debug logs only for that actor by adding this line to your `logback.xml`:

```
<logger name="fr.acinq.eclair.blockchain.bitcoind.zmq" level="DEBUG" />
```

What would really give us all the data needed would be to have zmq logs on the `bitcoind` side as well, to see on which side the issue is when that happens, but I don't know what `bitcoind` provides for that. I'll ask some core devs when I get the chance.

When the associated `bitcoind` node is down, there isn't much `eclair` can do.
Eclair must stay online, because it needs to continue processing lightning messages (e.g. settling htlcs to avoid channels closing).
It will run into issues if it needs to do on-chain operations, but it has no way out...
It can only notify the node operator that there is an issue with the `bitcoind` node (which it currently does with the blockchain watchdog events) but something needs to be done manually to restore the `bitcoind` node.

We could imagine more resilient setups, where you could have two bitcoin nodes (a primary and a secondary) and eclair would switch to the secondary node if the first one is down, but if both nodes are down this issue would still exist.

**Comment by engenegr** [comment]
@t-bast I've enabled ZMQ debug option but didn't see anything suspicious in logs so far. Just the same frequent disconnect-connect messages which may be related somehow to real network disruptions
```
2021-08-13 21:33:57,367 WARN  f.a.e.b.b.zmq.ZMQActor - disconnected from tcp://192.168.2.7:29000
2021-08-13 21:33:57,371 INFO  f.a.e.b.b.zmq.ZMQActor - connected to tcp://192.168.2.7:29000
2021-08-13 21:36:07,916 WARN  f.a.e.b.b.zmq.ZMQActor - disconnected from tcp://192.168.2.7:29000
2021-08-13 21:36:07,916 INFO  f.a.e.b.b.zmq.ZMQActor - connected to tcp://192.168.2.7:29000
```
For now I can't replicate the same error when ZMQ couldn't reconnect again, but I continue to observe node.

**Comment by engenegr** [comment]
Meanwhile I discovered the same error as in https://github.com/ACINQ/eclair/issues/935

```
2021-08-13 19:34:43,657 ERROR f.a.e.b.b.r.BatchingClient - got exception for batch of 1 requestsjava.net.SocketTimeoutException: timeout
	at okio.SocketAsyncTimeout.newTimeoutException(Okio.kt:159)
	at okio.AsyncTimeout.exit$jvm(AsyncTimeout.kt:203)
	at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:163)
	at okio.RealBufferedSource.read(RealBufferedSource.kt:41)
	at okhttp3.internal.http1.Http1ExchangeCodec$AbstractSource.read(Http1ExchangeCodec.kt:352)
	at okhttp3.internal.http1.Http1ExchangeCodec$FixedLengthSource.read(Http1ExchangeCodec.kt:389)
	at okhttp3.internal.connection.Exchange$ResponseBodySource.read(Exchange.kt:279)
	at okio.Buffer.writeAll(Buffer.kt:1104)
	at okio.RealBufferedSource.readString(RealBufferedSource.kt:194)
	at com.softwaremill.sttp.okhttp.OkHttpBackend$$anon$2.$anonfun$handleBasic$4(OkHttpBackend.scala:125)
	at scala.util.Try$.apply(Try.scala:210)
	at com.softwaremill.sttp.okhttp.OkHttpBackend$$anon$2.handleBasic(OkHttpBackend.scala:125)
	at com.softwaremill.sttp.ResponseAs$EagerResponseHandler.handle(ResponseAs.scala:67)
	at com.softwaremill.sttp.ResponseAs$EagerResponseHandler.handle$(ResponseAs.scala:63)
	at com.softwaremill.sttp.okhttp.OkHttpBackend$$anon$2.handle(OkHttpBackend.scala:116)
	at com.softwaremill.sttp.okhttp.OkHttpBackend.readResponse(OkHttpBackend.scala:107)
	at com.softwaremill.sttp.okhttp.OkHttpAsyncBackend$$anon$3.onResponse(OkHttpBackend.scala:231)
	at okhttp3.RealCall$AsyncCall.run(RealCall.kt:138)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.net.SocketException: Socket closed
	at java.base/java.net.SocketInputStream.read(SocketInputStream.java:183)
	at java.base/java.net.SocketInputStream.read(SocketInputStream.java:140)
	at okio.InputStreamSource.read(Okio.kt:102)
	at okio.AsyncTimeout$source$1.read(AsyncTimeout.kt:159)
	... 18 common frames omitted
```

As I see it is exactly between reconnects of ZMQ actor.

**Comment by engenegr** [comment]
It happens now in real time. Eclair 4 blocks late behind Core node.

![image](https://user-images.githubusercontent.com/15425710/129531200-552c6480-8053-475f-9389-a934cca780cf.png)

`getinfo`
```
  "version": "0.6.2-SNAPSHOT-cafaeed",
  "features": {
    "activated": {
      "gossip_queries": "optional",
      "option_shutdown_anysegwit": "optional",
      "payment_secret": "mandatory",
      "option_data_loss_protect": "optional",
      "option_static_remotekey": "optional",
      "basic_mpp": "optional",
      "gossip_queries_ex": "optional",
      "option_support_large_channel": "optional",
      "var_onion_optin": "mandatory"
    },
    "unknown": []
  },
  "chainHash": "6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000",
  "network": "mainnet",
  "blockHeight": 696032,
```
Core `getblockchaininfo`

```
{"result":{"chain":"main","blocks":696036,"headers":696036,
```
While ZMQ receives transactions in logs.

**Comment by t-bast** [comment]
> Meanwhile I discovered the same error as in #935

This didn't crash your node, did it?
It's expected that you see these logs if your connection between eclair and bitcoind is having issues, as long as it doesn't crash your node it's ok (and we simply can't avoid it, if the connection doesn't work, RPC calls will fail).

> UPD. 4 blocks finally received by Eclair.

Ok, so it looks like you're only experiencing delays.
If your network connection between eclair and bitcoind is having issues, that seems expected, isn't it?
I'll try to investigate what logs we can have on the bitcoind side when I have a bit of free time, I think that's what we're missing to completely understand the issue.

IIRC you mentioned using a wireguard proxy as well, right? Does that have logs or metrics to let you know whether packets were dropped or connection issues were found?

**Comment by engenegr** [comment]
> If your network connection between eclair and bitcoind is having issues, that seems expected, isn't it?

It wasn't about network connection it was about how eclair receives blocks. Because I was pinging node from Eclair and it showed be 696036 blocks while eclair was being 4 blocks late.

**Comment by t-bast** [comment]
But here it's a sender issue, right? The issue is that your bitcoind node doesn't send the blocks to eclair, isn't it? From the ZMQ point of view, eclair cannot do anything about this, it's bitcoind that should send these blocks faster.

But what you're advocating is that we shouldn't rely on ZMQ alone, and actively request blocks from `bitcoind` regularly (e.g. every 15 minutes), is that right?

**Comment by engenegr** [comment]
I can't say for sure whether it is sender issue or eclair. I just see that Eclair does receive transactions via ZMQ thus Core does send them. 

> But what you're advocating is that we shouldn't rely on ZMQ alone, and actively request blocks from bitcoind regularly (e.g. every 15 minutes), is that right?

You already said that this is not really good idea from software architecture point of view. I understand that. Moreover unavailable ZMQ also means unavailable Core API. So the only alternative might be in utilizing Core API services like Blockcypher or public explorers working as backups for Core ZMQ. I have no recommendations, I know little Eclair code and have no good idea about that.

**Comment by engenegr** [comment]
Okay. Besides of original situation when ZMQ couldn't get up again I will also try to use my Python script to see if there is network issue, i.e. compare timestamps and how both clients receive blocks on the same machine.

**Comment by t-bast** [comment]
> You already said that this is not really good idea from software architecture point of view. I understand that.

Yes, I'd like to avoid it and instead have a reliable ZMQ mechanism.
But as you're experiencing issues with it, we need to figure out whether ZMQ is reliable enough or if we must have a fallback mechanism.

> I can't say for sure whether it is sender issue or eclair. I just see that Eclair does receive transactions via ZMQ thus Core does send them.

I'm wondering if it could be somehow linked to watermarks and buffer issues (as bitcoind's sample code sets it to `0` and I don't know if that's the default or not: https://github.com/bitcoin/bitcoin/blob/42b00a37580a11fbb81a3748d9e834ae8913cdba/contrib/zmq/zmq_sub.py#L46). I've mimicked `bitcoind`'s sample code [here](https://github.com/ACINQ/eclair/pull/1910), can you apply this commit to your node and see if that's better?

If I understand correctly how this works (but I'm not sure I do), ZMQ makes no guarantee on message delivery and some messages may be dropped from OS network buffers without any warning anyway (and since blocks are quite big messages, that could be what happens). I'm not sure yet how to check the configuration on these OS buffers and monitor if packets are evicted (and raise limits if possible).

**Comment by engenegr** [comment]
> I'm wondering if it could be somehow linked to watermarks and buffer issues

I've built this commit and I will try it.

As I said I tried to track ZMQ independently with Python. You can see on this Kamon plot last event in 27 min and then no events.
![image](https://user-images.githubusercontent.com/15425710/129605370-640694e1-85ba-4b9c-9622-e95fe9da7e41.png)
I was able to catch this moment when eclair is 2 blocks late in `getinfo` while it seems it already received the block in according to Kamon. Running Python script in a while haven't shown any bugs. Only about persistent half a second late if compare with Python's timestamps. So maybe these are issues related to some other actor or subsystem which verifies blocks?

**Comment by t-bast** [comment]
> So maybe these are issues related to some other actor or subsystem which verifies blocks?

That delay can definitely be explained, the `ZmqWatcher` actor is the actor that receives blocks from the zmq socket and then broadcasts it to the rest of the system. It starts by analyzing the block to check whether some of your channels may be affected by txs confirmed in that block, if your network connection to `bitcoind` is slow it may take some time to complete because this process makes many additional calls to `bitcoind`.

**Comment by engenegr** [comment]
This is a plot with two more lines in ZMQ setup. It seems events are more uniformly spaced if compared to previous plots. Maybe not ideal tool for analysis but I see it is relevant. What do you think @t-bast ?

![zmq-updated](https://user-images.githubusercontent.com/15425710/130045524-9056245c-b012-4187-b756-cd2862a7dbe8.png)

**Comment by t-bast** [comment]
I think it shows that blocks are received regularly enough, aren't they?
Note that there is also an artificial 2-seconds delay before we analyze the block: https://github.com/ACINQ/eclair/blob/9a0fc1471031c75cae1b5964c49deb9c56a25931/eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/ZmqWatcher.scala#L236

This would explain the delay between the python script and eclair.

**Comment by engenegr** [comment]
I'd say that before there were more shorter intervals. Sometimes it is hard to say precisely because before even plot aggregated several blocks in one event. Now what I see, node lags max 2 blocks. It may be explained by shorter intervals since now hashrate is growing.

**Comment by engenegr** [comment]
@t-bast 
My node experiences lag about 5 blocks max again. It started since last evening when I noticed 2 blocks lag. The difference to previous ZMQ code (without IDLE parameter) is that now I can see `received blockid` in logs and it gives an idea that ZMQ should be working correctly. 

I have no ideas so far about the nature of these issues. Maybe it is actually Core slowed down for same reason. Because `curl` request to my node needs more time which wasn't the case when it was freshly restarted.

**Comment by engenegr** [comment]
@t-bast another thing. While writing previous message I noticed that Kamon indicated event in 45 min. Now it is 50 and Eclair is on the same height 697337. I mean that the height was 697337 during all time I'm analysing and reporting its behaviour. 

![image](https://user-images.githubusercontent.com/15425710/130578318-98b2de0b-901b-4b64-a078-6079a9011d44.png)

Another thing is that what I see in logs
```
2021-08-24 07:15:35,356 DEBUG f.a.e.b.b.zmq.ZMQActor - received blockid=000000000000000000033950845478efacf30e649e28b00910d5a86580f68a84
2021-08-24 07:40:57,567 DEBUG f.a.e.b.b.zmq.ZMQActor - received blockid=00000000000000000004e1ee7ffa5470eb0951c0a259bfb587d4d0df77dd5146
```
ends with block 697330 which is different to current height 697337 i see in `getinfo`. And also there is no log output about event received in 45 min.

**Comment by engenegr** [comment]
That's for sure something related to Eclair.

```
2021-08-24 07:55:02,944 DEBUG f.a.e.b.b.zmq.ZMQActor - received blockid=0000000000000000000474f0525dfe47fb67cd7b48d1bad322a2c57ac005cbeb
```
this is 697331. I see 697337 in getinfo.

**Comment by engenegr** [comment]
I've restarted fully synced Core and it didn't help. It means that ZMQ works fine (before it could resend some blocks or something) but the problem somewhere else.

**Comment by t-bast** [comment]
I'm a bit lost in all your messages and not sure what to conclude. Now that you have the debug logs that let you know when a block is received on the ZMQ socket, are you receiving the blocks on time there or not?

If you're seeing the blocks come at approximately the right time in the `ZMQActor` logs, it's good news, it means the issue is in the processing phase and we'll be able to easily figure it out.

If you're not seeing the blocks in the `ZMQActor` logs, this is still a ZMQ issue.

**Comment by engenegr** [comment]
> receiving the blocks on time there or not

I've seen significant slow down of receiving blocks from what I've seen from Eclair ZMQ blocks. Moreover, its numbers were inconsistent with `getinfo` tip block number. It wasn't showing at least 6 or 7 blocks it should already receive in according to `geinfo`.

After restarting Bitcoind it received global network tip block but somehow missed 7 blocks between last received in according to ZMQ and current Core's tip block.

I am sorry for confusing report. But I dumped it all here also for myself so now I can explain it a bit clearer (I hope).

**Comment by t-bast** [comment]
Just to clarify, when you say `getinfo` you mean eclair's `getinfo` API, not `bitcoind`'s `getblockchaininfo`?

> Moreover, its numbers were inconsistent with getinfo tip block number. 

Can you detail that part? That's very surprising.

`getinfo` returns the value stored in `nodeParams.currentBlockHeight`, and this value is only updated when a block is received through zmq, so it's impossible that this matches a block that hasn't been received. The only possibility is that `getinfo` shows something **older** that what you see is received on the `ZMQActor`: in that case it means there's an issue in block processing and it's an important information for the investigation.

> I've seen significant slow down of receiving blocks from what I've seen from Eclair ZMQ blocks.

You mean that the blocks aren't received in time by the `ZMQActor`, right?

I really think that we need to get logs from the `bitcoind` side to know if these blocks are sent on the ZMQ socket or not, this may be purely a `bitcoind` issue when blocks are delayed when `bitcoind` is overloaded. I haven't had a chance to discuss this with bitcoin devs yet and couldn't find anything in the `bitcoind` documentation...

**Comment by engenegr** [comment]
> Just to clarify, when you say getinfo you mean eclair's getinfo API, not bitcoind's getblockchaininfo?

Yes.

> That's very surprising.

Damn, yes. I couldn't believe I see this.

>  it's impossible that this matches a block that hasn't been received

I assume so. But I could see in logs only information which was put in logs. Are you sure logs are printed out exactly in corresponding moment in code when block is received?

> You mean that the blocks aren't received in time by the ZMQActor, right?

Yes. I wrote earlier "ends with block 697330 which is different to current height 697337 i see in getinfo".

**Comment by t-bast** [comment]
Actually I was wrong, this can happen and is likely to happen in case the blocks we receive on the ZMQ socket aren't the last blocks, because what we do is:

1. Receive a block on the ZMQ socket in the `ZmqActor`
2. Send this block for processing to the `ZmqWatcher` actor
3. The `ZmqWacther` then calls `bitcoind`'s `getblockcount` RPC to update the internal block count and makes many calls to `bitcoind` to check if some of the transactions we're watching have been confirmed

So the block height that eclair memorizes will match `bitcoind`'s latest tip, regardless of which blocks are received via ZMQ.
That means we only need to receive *some* blocks on the ZMQ socket, not necessarily the last one.

Looking at this code in more details, it would actually be very simple to add a 30 minutes timer that resets every time we receive a block, and triggers the validation logic if no blocks are received in 30 minutes, which would protect against unreliable ZMQ sockets. I wasn't a big fan of this earlier because I thought it would require more code, but it may actually be much simpler than I thought, so I'll give that a try.

**Comment by btcontract** [comment]
Btw, is block itself needed or do we only need a new block event (in which case block hash would suffice)? 

Looking at `ZmqWatcher` and I don't see stuff like scanning block contents, rather everything seems to be done by calling bitcoind api. Am I missing something?

If not then switching to block hashes can save traffic for setups with remote bitcoind.

**Comment by t-bast** [comment]
That's also a good point, we only need an event and we're only using the block hash (and just for logging purposes).
We initially scanned the actual block, but that changed a while ago.
I'll prototype migrating to block hash to save bandwidth.

**Comment by t-bast** [comment]
Done in #1910 if you want to experiment with that on your node.

**Comment by engenegr** [comment]
@t-bast thanks. for block headers implementation. it will reduce amount of data I like it very much in my setup.

**Comment by engenegr** [comment]
I'd test it in `master` and together with enabled ZMQ Idle option.
https://github.com/ACINQ/eclair/pull/1917/files
